newrelic/api/opentelemetry.py

```diff
@@ -0,0 +1,490 @@
+# Tracer: we can think of this as the transaction.
+# Span: we can think of this as the trace.
+# Links functionality has now been enabled but not implemented yet.  Links are relationships
+#   between spans, but lateral in hierarchy.  In NR we only have parent-child relationships.
+#   We may want to preserve this information with a custom attribute.  We can also add this
+#   as a new attribute in a trace, but it will still not be seen in the UI other than a trace
+#   attribute.
+
+
+class Span(otel_api_trace.Span):
+    def __init__(
+        self,
+        name=None,
+        parent=None,  # SpanContext
+        resource=None,
+        attributes=None,
+        kind=otel_api_trace.SpanKind.INTERNAL,
+        nr_transaction=None,
+        nr_trace_type=FunctionTrace,
+        instrumenting_module=None,
+        *args,
+        **kwargs,
+    ):
+        self.name = name
+        self.otel_parent = parent
+        self.attributes = attributes or {}
+        self.kind = kind
+        self.nr_transaction = (
+            nr_transaction or current_transaction()
+        )  # This attribute is purely to prevent garbage collection
+        self.nr_trace = None
+        self.instrumenting_module = instrumenting_module
+
+        self.nr_parent = None
+        current_nr_trace = current_trace()
+        if (
+            not self.otel_parent
+            or (self.otel_parent and self.otel_parent.span_id == int(current_nr_trace.guid, 16))
+            or (self.otel_parent and isinstance(current_nr_trace, Sentinel))
+        ):
+            # Expected to come here if one of three scenarios have occured:
+            # 1. `start_as_current_span` was used.
+            # 2. `start_span` was used and the current span was explicitly set
+            #    to the newly created one.
+            # 3. Only a Sentinel Trace exists so far while still having a
+            #    remote parent. From OTel's end, this will be represented
+            #    as a `NonRecordingSpan` (and be seen as `None` at this
+            #    point). This covers cases where span is remote.
+            self.nr_parent = current_nr_trace
+        else:
+            # Not sure if there is a usecase where we could get in here
+            # but for debug purposes, we will raise an error
+            _logger.debug("Otel span and NR trace do not match nor correspond to a remote span")
+            _logger.debug("otel span: %s\nnewrelic trace: %s", self.otel_parent, current_nr_trace)
+            raise ValueError("Unexpected span parent scenario encountered")
+
+        if nr_trace_type == FunctionTrace:
+            trace_kwargs = {"name": self.name, "params": self.attributes, "parent": self.nr_parent}
+            self.nr_trace = nr_trace_type(**trace_kwargs)
+        elif nr_trace_type == DatastoreTrace:
+            trace_kwargs = {
+                "product": self.instrumenting_module,
+                "target": None,
```

Comment:
It looks like `target` maps to OTEL's `db.name`.

Reply:
It does!  The issue with the OTel attributes is that the `db.name` attribute is usually not set during the creation of the span, but rather later on in the span's operation.